drivers/segger/console_rtt.c

```diff
@@ -0,0 +1,552 @@
+ *   Return true if the receive fifo is not empty
+ *
+ ****************************************************************************/
+
+static bool rtt_rxavailable(FAR struct uart_dev_s *dev)
+{
+  struct rtt_serial_dev_s *priv = dev->priv;
+
+  return !!SEGGER_RTT_HasData(priv->channel);
+}
+
+/****************************************************************************
+ * Name: rtt_send
+ *
+ * Description:
+ *   This method will send one byte on the UART
+ *
+ ****************************************************************************/
+
+static void rtt_send(FAR struct uart_dev_s *dev, int ch)
+{
+  struct rtt_serial_dev_s *priv = dev->priv;
+
+  /* For console device */
+
+  SEGGER_RTT_PutChar(priv->channel, ch);
+}
+
+/****************************************************************************
+ * Name: rtt_txint
+ *
+ * Description:
+ *   Call to enable or disable TX interrupts
+ *
+ ****************************************************************************/
+
+static void rtt_txint(FAR struct uart_dev_s *dev, bool enable)
+{
+  if (enable)
+    {
+      uart_xmitchars(dev);
+    }
+}
+
+/****************************************************************************
+ * Name: uart_txready
+ *
+ * Description:
+ *   Return true if the tranmsit fifo is not full
+ *
+ ****************************************************************************/
+
+static bool rtt_txready(FAR struct uart_dev_s *dev)
+{
+  return true;
+}
+
+/****************************************************************************
+ * Name: rtt_txempty
+ *
+ * Description:
+ *   Return true if the transmit fifo is empty
+ *
+ ****************************************************************************/
+
+static bool rtt_txempty(FAR struct uart_dev_s *dev)
+{
+  return true;
+}
+
+/****************************************************************************
+ * Name: up_uartloop
+ ****************************************************************************/
+
+static int rtt_uartloop(int argc, FAR char **argv)
+{
+  while (1)
+    {
+      uart_recvchars(&g_rtt_console);
+      usleep(USEC_PER_TICK);
+    }
+
+  return 0;
+}
+
+#endif /* CONFIG_RTT_CONSOLE */
+
+#ifdef CONFIG_RTT_SERIAL
+
+/****************************************************************************
+ * Name: rtt_serial_ioctl
+ ****************************************************************************/
+
+static int rtt_serial_open(FAR struct file *filep)
+{
+  return 0;
+}
+
+/****************************************************************************
+ * Name: rtt_serial_ioctl
+ ****************************************************************************/
+
+static int rtt_serial_close(FAR struct file *filep)
+{
+  return 0;
+}
+
+/****************************************************************************
+ * Name: rtt_serial_read
+ ****************************************************************************/
+
+static ssize_t rtt_serial_read(FAR struct file *filep, FAR char *buffer,
```

Comment:
why not share the same code base with console

Reply:
RTT has a buffer inside, and the serial framework also needs to define a buffer, and more buffers will be needed
Performance is better when separated, and requires less memory